internal/locate/replica_selector.go

```diff
@@ -0,0 +1,535 @@
+			maxScoreIdxes = append(maxScoreIdxes, i)
+		}
+	}
+	if reloadRegion {
+		selector.region.setSyncFlags(needDelayedReloadPending)
+	}
+	if len(maxScoreIdxes) == 1 {
+		idx := maxScoreIdxes[0]
+		return replicas[idx]
+	} else if len(maxScoreIdxes) > 1 {
+		// if there are more than one replica with the same max score, we will randomly select one
+		// todo: consider use store statistics information to select a faster one.
+		idx := maxScoreIdxes[randIntn(len(maxScoreIdxes))]
+		return replicas[idx]
+	}
+	if s.busyThreshold > 0 {
+		// when can't find an idle replica, no need to invalidate region.
+		return nil
+	}
+	// when meet deadline exceeded error, do fast retry without invalidate region cache.
+	if !hasDeadlineExceededError(selector.replicas) {
+		selector.invalidateRegion()
+	}
+	metrics.TiKVReplicaSelectorFailureCounter.WithLabelValues("exhausted").Inc()
+	return nil
+}
+
+func (s *ReplicaSelectMixedStrategy) isCandidate(r *replica, isLeader bool, epochStale bool, liveness livenessState) bool {
+	if epochStale || liveness == unreachable {
+		// the replica is not available, skip it.
+		return false
+	}
+	maxAttempt := 1
+	if r.dataIsNotReady && !isLeader {
+		// If the replica is failed by data not ready with stale read, we can retry it with replica-read.
+		// 	after https://github.com/tikv/tikv/pull/15726, the leader will not return DataIsNotReady error,
+		//	then no need to retry leader again, if you try it again, you may got a NotLeader error.
+		maxAttempt = 2
+	}
+	if r.isExhausted(maxAttempt, 0) {
+		// attempts is exhausted, skip it.
+		return false
+	}
+	if s.leaderOnly && !isLeader {
+		return false
+	}
+	if s.busyThreshold > 0 && (r.store.EstimatedWaitTime() > s.busyThreshold || r.serverIsBusy || isLeader) {
+		return false
+	}
+	return true
+}
+
+const (
+	// The definition of the score is:
+	// MSB                                                                               LSB
+	// [unused bits][1 bit: LabelMatches][1 bit: PreferLeader][2 bits: NormalPeer + NotSlow]
+	flagLabelMatches = 1 << 3
+	flagPreferLeader = 1 << 2
+	flagNormalPeer   = 1
+	flagNotSlow      = 1
+)
+
+// calculateScore calculates the score of the replica.
+func (s *ReplicaSelectMixedStrategy) calculateScore(r *replica, isLeader bool) int {
+	score := 0
+	if isLeader {
+		if s.preferLeader {
+			score |= flagPreferLeader
+		} else if s.tryLeader {
+			if len(s.labels) > 0 {
+				// When the leader has matching labels, prefer leader than other mismatching peers.
+				score |= flagPreferLeader
+			} else {
+				score |= flagNormalPeer
+			}
+		}
+	} else {
+		if s.learnerOnly {
+			if r.peer.Role == metapb.PeerRole_Learner {
+				score |= flagNormalPeer
+			}
+		} else {
+			score |= flagNormalPeer
+		}
+	}
+	if r.store.IsStoreMatch(s.stores) && r.store.IsLabelsMatch(s.labels) {
+		score |= flagLabelMatches
+	}
+	if !r.store.healthStatus.IsSlow() {
+		score += flagNotSlow
+	}
+	if score > 0 && r.attempts > 0 {
+		score = score - 1
+	}
```

Comment:
I just noticed that how about making this rule a new bit (at the lowest position, and shift all other bits right)?
So that we can still take the benifit that a score value's bits can mostly show how it's calculated.

Btw consider adding some comment at line 341 to state that flagNotSlow and and flagNormalPeer is at the same position and 2 bits are reserved for them so that they can be added together.

Reply:
Nice catch, I use following score flag now:

```go
	// The definition of the score is:
	// MSB                                                                               LSB
	// [unused bits][1 bit: LabelMatches][1 bit: PreferLeader][2 bits: NormalPeer + NotSlow]
	flagLabelMatches = 1 << 4
	flagPreferLeader = 1 << 3
	flagNormalPeer   = 1 << 2
	flagNotSlow      = 1 << 1
	flagNotAttempt   = 1
```

Reply:
In fact, this is rare, usually each replica will only try 1 time, unless the replica has previously returned a dataNotReady error, and then there is a chance to try the replica again.